File: prolog/two_sum.pl
list_with_index(List, Res) :-
    list_with_index(List, 0, Res).

list_with_index([], _, []).

list_with_index([H|T], N, [num(H, N)|Res]) :-
    N1 is N + 1,
    list_with_index(T, N1, Res).

select(H, [H|T], T).
select(X, [H|T], [H|Res]) :-
    select(X, T, Res).

two_sum(List, Target, [M1, M2]):-
    list_with_index(List, NewList),
    select(num(E1, M1), NewList, _),
    select(num(E2, M2), NewList, _),
    Target is E1 + E2,
    M1 \= M2,
    !.

File: python/HashTable/twoSum/two_sum.py
def two_sum(nums, target):
    hash_table = {}

    for idx, num in enumerate(nums):
        if num not in hash_table:
            hash_table[num] = idx
        num1 = target - num
        if num1 in hash_table and hash_table[num1] != idx:
            return [hash_table[num1], idx]
    return False

print(two_sum([3, 2, 4], 6))

